datashader/utils.py

```diff
@@ -361,6 +361,8 @@ def dshape_from_pandas_helper(col):
     dataframe.
     """
     if isinstance(col.dtype, type(pd.Categorical.dtype)) or isinstance(col.dtype, pd.api.types.CategoricalDtype):
+        if not getattr(col.cat, 'known', True):
+            col = col.cat.as_known()
```

Comment:
This is potentially expensive, and since you're dropping the reference this is work that is lost. I'm not sure if there's a better way, users interested in performance should provide dataframes with known categories in beforehand.

Reply:
>This is potentially expensive, and since you're dropping the reference this is work that is lost.

Right I considered assigning back to the original dataframe but that would be a bit too magical. I think if we can get memoization to work I won't feel so bad about this. 

While I've got you here, I remember you telling me that calling ``.head()`` was not usually sufficient for inferring the categories and that only fastparquet loads and stores the categories in metadata. Having played around with loading a variety of different datasets from csv, parquet loaded via pyarrow and simply converting a pandas dataframe to a dask dataframe, the `.head()` seems to have consistently returned all available categories even if they are not present in `.head()`. Is that just a quirk of the examples I've tried or can this be relied on afterall?

Reply:
That's just  quirk of the examples you've tried.

```
In [38]: df = pd.DataFrame({'a': ['a', 'b', 'b']})

In [39]: df2 = pd.DataFrame({'a': ['b', 'b', 'c']})

In [40]: ddf = dd.concat([df, df2])

In [41]: ddf2 = ddf.astype({'a': 'category'})

In [42]: ddf2
Out[42]:
Dask DataFrame Structure:
                               a
npartitions=2
               category[unknown]
                             ...
                             ...
Dask Name: astype, 6 tasks

In [43]: ddf2.head(1).a.cat.categories
Out[43]: Index(['a', 'b'], dtype='object')

In [44]: ddf2.a.cat.as_known().cat.categories
Out[44]: Index(['a', 'b', 'c'], dtype='object')
```

Reply:
Thanks!